In [1]:
#Overarching importation
import sys, os
PROJECT_ROOT = os.path.abspath("..")  # go up one directory from notebooks/
if PROJECT_ROOT not in sys.path:
    sys.path.insert(0, PROJECT_ROOT)

In [5]:
from src.Ingest_Data import DataIngestorFactory

#-------#
# To help with reading and manipulating data
import pandas as pd
import numpy as np

# To help with data visualization
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

# To be used for missing value imputation
from sklearn.impute import SimpleImputer

# To help with model building
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    AdaBoostClassifier,
    GradientBoostingClassifier,
    RandomForestClassifier,
    BaggingClassifier,
)
from xgboost import XGBClassifier

# To get different metric scores, and split data
from sklearn import metrics
from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.metrics import (
    f1_score,
    accuracy_score,
    recall_score,
    precision_score,
    confusion_matrix,
    roc_auc_score
)

# To oversample and undersample data
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler

# To be used for data scaling and one hot encoding
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder

# To be used for tuning the model
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

# To use statistical functions
import scipy.stats as stats

# To be used for creating pipelines and personalizing them
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# To define maximum number of columns to be displayed in a dataframe
pd.set_option("display.max_columns", None)

# To supress scientific notations for a dataframe
pd.set_option("display.float_format", lambda x: "%.3f" % x)

# To supress warnings
import warnings

warnings.filterwarnings("ignore")

In [6]:
#Load Data
BASE_URL = "http://127.0.0.1:8000"
TRAIN_ENDPOINT = f"{BASE_URL}/train"
TEST_ENDPOINT = f"{BASE_URL}/test"

In [7]:
DataIngestorFactory.get_data_ingestor(TRAIN_ENDPOINT).ingest(TRAIN_ENDPOINT)

2025-09-28 03:31:20 [INFO] src.Ingest_Data: Fetching data from API endpoint!!: http://127.0.0.1:8000/train
2025-09-28 03:31:23 [INFO] src.Ingest_Data: Received 40000 records from API


V1            V2           V3            V4            V5  \
0      -3.50521329   5.070816556  2.500694783   0.751565909   5.414091884   
1      3.840802303  -0.744551858  7.325602151  -5.815326648  -3.062697946   
2      0.909041707  -2.659715845   5.23223366  -1.308306897  -2.524055462   
3      5.540182389  -3.418732874  2.850160289  -4.977010325  -3.266263267   
4      4.931424537   1.703925804  0.378594124  -1.852290737  -1.982952197   
...            ...           ...          ...           ...           ...   
39995  1.175070685  -1.354790335  4.106679832   4.014808477   -3.08519116   
39996  3.775296069   3.824924852  2.867824637   3.368808378  -0.679749651   
39997   4.40251548   1.898080401   8.21322859  -5.058250262  -0.004670586   
39998  6.670838815   0.560989556  4.951878588  -8.838304191   0.115902667   
39999  -3.82735588  -1.322654134  3.270700181   4.096240525  -0.245822996   

                 V6            V7            V8            V9           V10  \
0       0.085034014  -2.304466047   0.118047857  -1.917490871  -1.481235507   
1       0.480068893   0.544013551  -4.043584393  -1.173479445   3.570352181   
2      -0.539079492  -2.809633119  -3.273907609   1.266268536   2.609635312   
3      -2.005542309   0.351839972   4.183686481  -1.042189162   0.005901722   
4       2.618002739  -0.037659152  -5.684056956   1.974835984   0.635661901   
...             ...           ...           ...           ...           ...   
39995    -2.8877693  -3.050924309  -1.069377322   3.245813942  -0.370393347   
39996   -0.93094318  -0.629292982  -2.947774279    2.73803133  -1.023083943   
39997    0.26511292   0.280303465  -4.859109781   0.583682242   2.709747088   
39998   2.060838281   0.828419718  -3.065245391  -1.719198788    2.69843508   
39999  -3.370795876  -2.261325458   0.928429515  -0.308876914    1.83169191   

                V11           V12          V13           V14           V15  \
0      -0.401580755  -0.383877161  2.093071066   -2.54405025  -6.322206646   
1       1.516160499   3.125625688  3.469452029  -1.517163959  -0.010414687   
2      -0.376448973  -1.266170538  5.949901722   0.193781385  -5.839702123   
3       0.511535252   4.027686478  3.567224259   1.548408543   0.968034206   
4       3.733109986  -1.529581199   2.08796597  -1.975366299  -2.540944126   
...             ...           ...          ...           ...           ...   
39995  -3.216258494  -1.043554297  4.296519355  -0.440718538  -5.827998491   
39996  -1.784042431  -1.500553958  0.729498711  -2.371941144  -2.435113753   
39997  -2.475479055   -0.52946207  2.965336937  -0.920127176  -0.169825241   
39998   0.114142177   0.843902412  0.875431051  -0.673725526   1.535266553   
39999  -3.735426852   1.036158038  3.113595481   0.438078199  -4.615141395   

                V16           V17           V18           V19           V20  \
0       0.101501597  -3.077733811    2.18948964   -2.76623031  -8.532590498   
1      -4.603568994  -7.747528181  -0.092048171   3.502156424  -4.092504943   
2       -6.11081602  -6.215282502    0.96225224   2.062659187  -0.235606345   
3       1.966342172  -2.315081966  -0.445240945  -0.655340221   1.105057898   
4      -7.336932177   -5.21599569   0.914929596   1.820737733  -5.433117669   
...             ...           ...           ...           ...           ...   
39995  -7.480750973   0.869538473   -0.12481826   4.162417932   5.316583601   
39996  -7.403036414   0.845352509  -0.654052754   3.320158029  -1.116302402   
39997  -5.557410998  -6.071037355   0.686798156  -0.582803849  -5.006768701   
39998  -3.481283381   -6.84135868    2.92660457  -2.672359903  -8.330404239   
39999   -1.82477945  -0.097874555  -0.817420911    4.14924135   3.231469076   

                V21           V22           V23           V24           V25  \
0      -4.035947699  -0.451932017   2.344674807   8.755114487   -5.63217525   
1      -5.896325959    0.18360786  -6.979060653  -4.212251559   2.162944114   
2      -7.924292746

## Data Overview

In [ ]:
df_train

In [ ]:
df_train.tail()

In [ ]:
df_train.info()

In [ ]:
# let's check for duplicate values in the data
df_train.duplicated().sum()

In [ ]:
df_train.nunique()

## Univariate Analysis

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import math

def plot_boxplots(df, cols_per_page=12, n_cols=3, figsize=(15, 20)):
    cols = df.columns.tolist()
    n = len(cols)
    n_pages = math.ceil(n / cols_per_page)

    for page in range(n_pages):
        start = page * cols_per_page
        end = start + cols_per_page
        subset = cols[start:end]

        n_rows = math.ceil(len(subset) / n_cols)

        fig, axes = plt.subplots(n_rows, n_cols, figsize=figsize)
        axes = axes.flatten()

        for i, variable in enumerate(subset):
            sns.boxplot(x=df[variable], ax=axes[i])
            axes[i].set_title(variable)

        # Hide any unused subplots
        for j in range(i+1, len(axes)):
            axes[j].set_visible(False)

        plt.tight_layout()
        plt.show()


In [ ]:
plot_boxplots(df_train, cols_per_page=12, n_cols=3, figsize=(15, 12))

In [ ]:
plt.figure(figsize=(6, 4))
sns.countplot(df_train["Target"],)
plt.show()
df_train["Target"].value_counts()

- "Target" class is imbalanced with 37802 or 94.5% "No failures (i.e., 0)" and 2198 or 5.4% "Failures (i.e., 1)"

## Data Pre-processing
- Missing value Handing
- 